contents/security/passwordStorage.md

```diff
@@ -0,0 +1,120 @@
+
+One question that is commonly asked by developers is where to store the salt. The salt can be stored in plaintext, along with the user in the database. Since the goal of the salt is only to prevent precomputed rainbow tables from being used, it doesn't need to be encrypted in the database.
+
+### If an attacker has gained access to the entire application, what does it matter if my passwords are stored in plaintext or not?
+
+If an attacker has already gained access to the entire application, then he already has all the information that he could possibly want from your server. He would have access to all of your application's data, including data from your users or from any analytics software that you might be running. However, by adding salt and hashing your passwords, the attacker still doesn't know your customer's passwords and could take years to find out. Otherwise, since 59% of people use the same password across multiple sites <sup>[source](https://securityboulevard.com/2018/05/59-of-people-use-the-same-password-everywhere-poll-finds/)</sup>, the attacker could quickly try other websites such as banks to attempt to break into those accounts, which can yield great financial return.
+
+Additionally, when a user signed up on your website and provided you a password, they implicitly trusted you to keep that information safe and secure for them. In a sense, you have a responsibility to keep their passwords secure as well. By doing proper password storage, if your servers ever get breached, you can assure your customers that their passwords are properly secured and maintain some of their trust in you.
+
+## Getting started
+
+<p style="color: #721c24; background-color: #f8d7da; border-color: #f5c6cb; padding: .75rem 1.25rem; border: 1px solid transparent; border-radius: .25rem;">
+Reading this article doesn't make you an expert in writing your own password hashing code. It is far too easy to screw up and make a mistake. Instead, use one of the free libraries that provide a crypto function that has already been well-tested by the community. <b>Do not write your own crypto library.</b>
+</p>
+
+Here are some libraries you can use to implement password storage, depending on your language:
+* If you're using PHP, the [password_hash](https://secure.php.net/manual/en/function.password-hash.php) function will do the job for you. 
+* If you're using java, use [SecretKeyFactory](https://www.owasp.org/index.php/Hashing_Java) to apply hashes with SHA256.
+* If you're using Python, use [hashlib](https://docs.python.org/3/library/hashlib.html) to salt and hash your password. [Example usage](https://stackoverflow.com/questions/9594125/salt-and-hash-a-password-in-python)
```

Comment:
[`passlib`](https://passlib.readthedocs.io/en/stable/) may be a better recommendation for this.